### PR TITLE
20108 content library tags improvements

### PR DESCRIPTION
**Pull request by DudkevichD**
https://trello.com/c/hlori35i/20108-content-library-tags-improvements

### CODE REVIEW

**Comment by DudkevichD** [comment]
image

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 1]
```diff
@@ -3,25 +3,31 @@ import material/material_ui;
...
 
-	MInputChipsStyle ::= MInputChipsMaxWidth, MSelectedInputChips, MInputChipsLeftCaption, MCustomAutoCompleteStyle, MCustomChipStyle,
-		MInputChipsMultiline, MAnimationEnabled /*for multiline*/;
+	MInputChipsStyle ::= MInputChipsMaxWidth, MSelectedInputChips, MInputChipsLeftCaption, MCustomAutoCompleteStyle, MCustomChipStyle, MInputChipsTextButton,
+		MInputChipsToRight, MInputChipsInputTrigger, MInputChipsMultiline, MAnimationEnabled /*for multiline*/;
 
 		MSelectedInputChips(items : DynamicBehaviour<[string]>);
 		MInputChipsLeftCaption(caption : Material);
+		MInputChipsTextButton(text : string);
+		MInputChipsToRight(left : bool);
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:12]: I'd rename it to MInputChipsAlignRight and remove left parameter.

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 2]
```diff
@@ -3,25 +3,31 @@ import material/material_ui;
...
 		MInputChipsMaxWidth(width : Transform<double>);
 		MCustomChipStyle(style : [MChipStyle]);
 		MCustomAutoCompleteStyle(style : [MAutoCompleteStyle]);
 		MInputChipsMultiline();
+		MInputChipsInputTrigger(trigget : DynamicBehaviour<bool>);
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:17]: 'trigget' typo

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 3]
```diff
@@ -116,15 +137,19 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
 				TScrollAxisEnabled(const(true), const(false))
 			]);
 	
-	MBaselineCols2(caption, chips)
+	MBaselineCols([caption, chips])
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:140]: Why is it for?

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 4]
```diff
@@ -33,12 +39,26 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
...
+		)
+	}
 
 	autoComplete =
 		MGetManager(\manager -> {
 			defFn = defaultCompletionFn(manager, manager.focus, []).fn;
-			MSelect(items, \itms ->
-				MAutoComplete(text, itms, concatA([
+			MSelect2(items, selectedItems, \itms, selecItms ->
+				MAutoComplete(text, filter(itms,\i -> !contains(selecItms, i)), concatA([
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:61]: Don't filter dictionary here, results should be filtered inside MSentenceMatcher.

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 5]
```diff
@@ -93,7 +113,8 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
 	chipsGrid =
 		if (multiline) {
 			MFlexibleGrid(
-				farrayPush(fmap(selectedItems, materialize), const(autoComplete)),
+				if(chipsToRight) fconcat(fconcat(farray(const(MAvailable(autoComplete, MFillWY(240.)))), farray(make(button))), fmap(selectedItems, materialize))
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:116]: const(button) ?

File: lib/material/extra/input_chips/material_input_chips.flow (hunk 2 above)
Comment (NikitaFil) [reply lib/material/extra/input_chips/material_input_chips.flow:17]: Also, don't understand how is it meant to be used.

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 6]
```diff
@@ -116,15 +137,19 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
...
 	|> (\m -> if (multiline) MAttachWidth(m, fullWidth) else MBaselineCols2(m, autoComplete))
 	|> (\m -> MAttachAvailableWidth(m, avWidth))
 	|> (\m -> eitherMap(maxWidth, \mw -> MAvailable(m, TFillWYT(mw.width)), m))
 	|> (\m -> MFrame(0., 0., [MFill(MWhite())], m))
-	|> (\m -> MConstruct([makeSubscribe2(items, \itms ->
-			nextDistinct(selectedItems, filter(getValue(selectedItems), \it -> contains(itms, it)))
-		)], m))
+	|> (\m -> MConstruct(
+		[
+			makeSubscribe(selectedItems, \i -> nextDistinct(selectedItems, uniq(i))),
```
Comment (NikitaFil) [inline lib/material/extra/input_chips/material_input_chips.flow:148]: Better to update with delay. Also, add a style for this behaviour, it's not always an expected one.

**Review by NikitaFil — CHANGES_REQUESTED** [review]

File: lib/material/extra/input_chips/material_input_chips.flow (hunk 2 above)
Comment (DudkevichD) [reply lib/material/extra/input_chips/material_input_chips.flow:17]: Fixed to MInputChipsText(text : Dynamic Behaviour<string>);
to keep track of whether the string is empty

File: lib/material/extra/input_chips/material_input_chips.flow (hunk 4 above)
Comment (DudkevichD) [reply lib/material/extra/input_chips/material_input_chips.flow:61]: Now I am using MDictionaryDynamic(fselect2

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 7]
```diff
@@ -33,13 +41,28 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
...
+		)
+	}
 
 	autoComplete =
 		MGetManager(\manager -> {
 			defFn = defaultCompletionFn(manager, manager.focus, []).fn;
 			MSelect(items, \itms ->
 				MAutoComplete(text, itms, concatA([
 					[
+						MDictionaryDynamic(fselect2(items, selectedItems, FLift2(\it, s -> filter(it, \i-> !contains(s, i))))),
```
Comment (pzotov) [inline lib/material/extra/input_chips/material_input_chips.flow:65]: subtractA

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 8]
```diff
@@ -93,7 +116,8 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
 	chipsGrid =
 		if (multiline) {
 			MFlexibleGrid(
-				farrayPush(fmap(selectedItems, materialize), const(autoComplete)),
+				if(chipsToRight) fconcat(fconcat(farray(const(MAvailable(autoComplete, MFillWY(240.)))), farray(const(button))), fmap(selectedItems, materialize))
```
Comment (pzotov) [inline lib/material/extra/input_chips/material_input_chips.flow:119]: const([MAvailable(autoComplete, MFillWY(240.)), button])?

File: lib/material/extra/input_chips/material_input_chips.flow [hunk 9]
```diff
@@ -122,9 +146,15 @@ MInputChips(items : Transform<[string]>, style : [MInputChipsStyle]) -> Material
...
 	|> (\m -> MFrame(0., 0., [MFill(MWhite())], m))
-	|> (\m -> MConstruct([makeSubscribe2(items, \itms ->
-			nextDistinct(selectedItems, filter(getValue(selectedItems), \it -> contains(itms, it)))
-		)], m))
+	|> (\m -> MConstruct(
+		ifArrayPushMany(
+			[makeSubscribe2(items, \itms -> nextDistinct(selectedItems, filter(getValue(selectedItems), \it -> contains(itms, it))))],
+			[
+				Pair(uniqChips, makeSubscribe(selectedItems, \i -> deferred(\-> nextDistinct(selectedItems, uniq(i))))),
+				Pair(inputText.track, makeSubscribe(text, \t -> nextDistinct(inputText.text, t))),
```
Comment (pzotov) [inline lib/material/extra/input_chips/material_input_chips.flow:154]: fconnect()

**Review by pzotov — APPROVED** [review]
Looks good, but check small comments

**Pull request merged**